File: anchor/payloads.py
"""Dependency-free validation for the proposed §4b committed payloads."""

from __future__ import annotations

import json
from pathlib import Path
import re
from typing import Any

from anchor.canonical import canonical, canonicalize_json


_SCHEMA_DIR = Path(__file__).resolve().parents[1] / "contracts" / "payloads"
_SCHEMA_FILES = {
    "checkin_opened": "checkin_opened.v1.json",
    "checkin_result": "checkin_result.v1.json",
    "journey_ended": "journey_ended.v1.json",
}
_DIALECT = "https://json-schema.org/draft/2020-12/schema"
_VALIDATION_KEYWORDS = {
    "type",
    "const",
    "enum",
    "pattern",
    "minLength",
    "maxLength",
    "minimum",
    "required",
    "additionalProperties",
    "properties",
}
# These standard JSON Schema annotations carry documentation, not validation
# behaviour. They are explicitly accepted; every other keyword fails closed.
_ANNOTATION_KEYWORDS = {"$schema", "$id", "title", "description", "examples"}
_SUPPORTED_KEYWORDS = _VALIDATION_KEYWORDS | _ANNOTATION_KEYWORDS


class PayloadError(ValueError):
    """A payload is invalid or a schema uses unsupported semantics."""

    def __init__(self, reason: str):
        self.reason = reason
        super().__init__(reason)


def _same_json_value(left: Any, right: Any) -> bool:
    if type(left) is not type(right):
        return False
    if isinstance(left, dict):
        return left.keys() == right.keys() and all(
            _same_json_value(left[key], right[key]) for key in left
        )
    if isinstance(left, list):
        return len(left) == len(right) and all(
            _same_json_value(a, b) for a, b in zip(left, right)
        )
    return left == right


def _assert_supported_schema(schema: Any, path: str = "$schema") -> None:
    if not isinstance(schema, dict):
        raise PayloadError(f"schema at {path} must be an object")
    unsupported = sorted(set(schema) - _SUPPORTED_KEYWORDS)
    if unsupported:
        raise PayloadError(f"unsupported schema keyword {unsupported[0]} at {path}")
    if "$schema" in schema and schema["$schema"] != _DIALECT:
        raise PayloadError(f"unsupported schema dialect at {path}")
    if "additionalProperties" in schema and schema["additionalProperties"] is not False:
        raise PayloadError(f"unsupported additionalProperties value at {path}")
    properties = schema.get("properties")
    if properties is not None:
        if not isinstance(properties, dict):
            raise PayloadError(f"properties at {path} must be an object")
        for name, child in properties.items():
            if not isinstance(name, str):
                raise PayloadError(f"property name at {path} must be a string")
            _assert_supported_schema(child, f"{path}.properties.{name}")
    required = schema.get("required")
    if required is not None and (
        not isinstance(required, list)
        or any(not isinstance(name, str) for name in required)
    ):
        raise PayloadError(f"required at {path} must be an array of strings")
    for keyword in ("minLength", "maxLength"):
        value = schema.get(keyword)
        if value is not None and (type(value) is not int or value < 0):
            raise PayloadError(f"unsupported {keyword} value at {path}")
    minimum = schema.get("minimum")
    if minimum is not None and (type(minimum) is not int or abs(minimum) > 2**53 - 1):
        raise PayloadError(f"unsupported minimum value at {path}")
    if "pattern" in schema:
        if not isinstance(schema["pattern"], str):
            raise PayloadError(f"pattern at {path} must be a string")
        try:
            re.compile(schema["pattern"])
        except re.error as exc:
            raise PayloadError(f"invalid pattern at {path}") from exc
    if "enum" in schema and not isinstance(schema["enum"], list):
        raise PayloadError(f"enum at {path} must be an array")
    if "type" in schema and (
        not isinstance(schema["type"], str)
        or schema["type"] not in {"object", "string", "integer"}
    ):
        raise PayloadError(f"unsupported type keyword value at {path}")


def _fail(reason: str, path: str) -> None:
    raise PayloadError(f"{reason} at {path}")


def _validate_schema(payload: Any, schema: dict, path: str = "$payload") -> None:
    _assert_supported_schema(schema)

    expected_type = schema.get("type")
    if expected_type == "object" and not isinstance(payload, dict):
        _fail("must be an object", path)
    if expected_type == "string" and not isinstance(payload, str):
        _fail("must be a string", path)
    if expected_type == "integer" and type(payload) is not int:
        _fail("must be an integer", path)

    if "const" in schema and not _same_json_value(payload, schema["const"]):
        _fail(f"must equal {schema['const']!r}", path)
    if "enum" in schema and not any(
        _same_json_value(payload, candidate) for candidate in schema["enum"]
    ):
        _fail("is not an allowed value", path)

    if "pattern" in schema and isinstance(payload, str):
        if re.search(schema["pattern"], payload) is None:
            _fail("does not match the required pattern", path)
    if "minLength" in schema and isinstance(payload, str):
        if len(payload) < schema["minLength"]:
            _fail(f"must contain at least {schema['minLength']} code points", path)
    if "maxLength" in schema and isinstance(payload, str):
        if len(payload) > schema["maxLength"]:
            _fail(f"must contain at most {schema['maxLength']} code points", path)
    if "minimum" in schema and type(payload) is int:
        if payload < schema["minimum"]:
            _fail(f"must be at least {schema['minimum']}", path)

    if "required" in schema:
        if not isinstance(payload, dict):
            _fail("must be an object to satisfy required fields", path)
        for name in schema["required"]:
            if name not in payload:
                _fail(f"is missing required property {name!r}", path)

    if "properties" in schema:
        if not isinstance(payload, dict):
            _fail("must be an object to satisfy properties", path)
        for name, value in payload.items():
            property_schema = schema["properties"].get(name)
            if property_schema is None:
                if schema.get("additionalProperties") is False:
                    _fail(f"contains unsupported property {name!r}", path)
                continue
            _validate_schema(value, property_schema, f"{path}.{name}")

    if schema.get("additionalProperties") is False and "properties" not in schema:
        if not isinstance(payload, dict):
            _fail("must be an object for additionalProperties check", path)
        _fail("additionalProperties false requires properties", path)


def _load_schema(kind: str) -> dict:
    filename = _SCHEMA_FILES.get(kind)
    if filename is None:
        raise PayloadError(f"unsupported payload kind {kind!r}")
    try:
        schema = json.loads((_SCHEMA_DIR / filename).read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as exc:
        raise PayloadError(f"could not load schema for {kind!r}") from exc
    _assert_supported_schema(schema)
    return schema


def validate_payload(kind: str, payload: Any) -> Any:
    """Validate one parsed payload against its checked-in §4b schema.

    JSON Schema annotations are permitted as metadata. Any unrecognized
    validation keyword or unsupported schema value raises instead of being
    silently ignored.
    """
    schema = _load_schema(kind)
    return validate_schema(payload, schema)


def validate_schema(payload: Any, schema: dict) -> Any:
    """Validate against a supplied schema; exported for fail-closed tests."""
    try:
        canonical(payload)
    except (TypeError, ValueError) as exc:
        raise PayloadError(f"invalid canonical payload: {exc}") from exc
    _validate_schema(payload, schema)
    return payload


def validate_payload_json(kind: str, json_text: str) -> Any:
    """Validate raw JSON, rejecting duplicate keys/floats before parsing it."""
    if not isinstance(json_text, str):
        raise PayloadError("payload JSON must be text")
    try:
        canonicalize_json(json_text)
        payload = json.loads(json_text)
    except (ValueError, json.JSONDecodeError) as exc:
        raise PayloadError(f"invalid canonical payload JSON: {exc}") from exc
    return validate_payload(kind, payload)

File: shared/payloads.js
import pinAuthorisedSchema from "../contracts/payloads/pin_authorised.v1.json" with { type: "json" };
// Dependency-free validator for the proposed §4b payload schemas.
// Validation keywords outside this explicit subset fail closed. Static JSON
// imports keep this module usable in both Node tests and browser bundles.
import checkinOpenedSchema from "../contracts/payloads/checkin_opened.v1.json" with { type: "json" };
import checkinResultSchema from "../contracts/payloads/checkin_result.v1.json" with { type: "json" };
import journeyEndedSchema from "../contracts/payloads/journey_ended.v1.json" with { type: "json" };
import guardianAckSchema from "../contracts/payloads/guardian_ack.v1.json" with { type: "json" };
import evidenceObservedV1Schema from "../contracts/payloads/evidence_observed.v1.json" with { type: "json" };
import evidenceObservedV2Schema from "../contracts/payloads/evidence_observed.v2.json" with { type: "json" };
import { canonicalize, canonicalizeJson } from "./canonical.js";

const DIALECT = "https://json-schema.org/draft/2020-12/schema";
const VALIDATION_KEYWORDS = new Set([
  "type",
  "const",
  "enum",
  "pattern",
  "minLength",
  "maxLength",
  "minimum",
  "maximum",
  "maxItems",
  "items",
  "required",
  "additionalProperties",
  "properties",
]);
// Standard annotation keys are permitted as inert documentation metadata.
const ANNOTATION_KEYWORDS = new Set(["$schema", "$id", "title", "description", "examples"]);
const SUPPORTED_KEYWORDS = new Set([...VALIDATION_KEYWORDS, ...ANNOTATION_KEYWORDS]);
const SCHEMAS = Object.freeze({
  checkin_opened: checkinOpenedSchema,
  checkin_result: checkinResultSchema,
  journey_ended: journeyEndedSchema,
  guardian_ack: guardianAckSchema,
  pin_authorised: pinAuthorisedSchema,
  // Keyed by pv, as anchor/payloads.py does (ADR-0047, PROPOSED).
  evidence_observed: Object.freeze({ 1: evidenceObservedV1Schema, 2: evidenceObservedV2Schema }),
});
const TYPES = new Set(["object", "string", "integer", "array", "null"]);

export class PayloadError extends Error {
  constructor(reason) {
    super(reason);
    this.name = "PayloadError";
    this.reason = reason;
  }
}

function fail(reason, path) {
  throw new PayloadError(`${reason} at ${path}`);
}

function isObject(value) {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function sameJsonValue(left, right) {
  if (typeof left !== typeof right) return false;
  if (Array.isArray(left) || Array.isArray(right)) {
    return Array.isArray(left) && Array.isArray(right) && left.length === right.length &&
      left.every((value, index) => sameJsonValue(value, right[index]));
  }
  if (isObject(left) || isObject(right)) {
    if (!isObject(left) || !isObject(right)) return false;
    const leftKeys = Object.keys(left);
    const rightKeys = Object.keys(right);
    return leftKeys.length === rightKeys.length && leftKeys.every((key) =>
      Object.hasOwn(right, key) && sameJsonValue(left[key], right[key]));
  }
  return Object.is(left, right);
}

function assertSupportedSchema(schema, path = "$schema") {
  if (!isObject(schema)) throw new PayloadError(`schema at ${path} must be an object`);
  const unsupported = Object.keys(schema).filter((key) => !SUPPORTED_KEYWORDS.has(key)).sort();
  if (unsupported.length) {
    throw new PayloadError(`unsupported schema keyword ${unsupported[0]} at ${path}`);
  }
  if (Object.hasOwn(schema, "$schema") && schema.$schema !== DIALECT) {
    throw new PayloadError(`unsupported schema dialect at ${path}`);
  }
  if (Object.hasOwn(schema, "additionalProperties") && schema.additionalProperties !== false) {
    throw new PayloadError(`unsupported additionalProperties value at ${path}`);
  }
  if (Object.hasOwn(schema, "properties")) {
    if (!isObject(schema.properties)) throw new PayloadError(`properties at ${path} must be an object`);
    for (const [name, child] of Object.entries(schema.properties)) {
      assertSupportedSchema(child, `${path}.properties.${name}`);
    }
  }
  if (Object.hasOwn(schema, "required") &&
      (!Array.isArray(schema.required) || schema.required.some((name) => typeof name !== "string"))) {
    throw new PayloadError(`required at ${path} must be an array of strings`);
  }
  if (Object.hasOwn(schema, "maxItems") && (!Number.isSafeInteger(schema.maxItems) || schema.maxItems < 0)) {
    throw new PayloadError(`unsupported maxItems value at ${path}`);
  }
  if (Object.hasOwn(schema, "items")) assertSupportedSchema(schema.items, `${path}.items`);
  for (const keyword of ["minLength", "maxLength", "minimum", "maximum"]) {
    if (Object.hasOwn(schema, keyword) &&
        (!Number.isSafeInteger(schema[keyword]) ||
          (keyword !== "minimum" && keyword !== "maximum" && schema[keyword] < 0))) {
      throw new PayloadError(`unsupported ${keyword} value at ${path}`);
    }
  }
  if (Object.hasOwn(schema, "pattern")) {
    if (typeof schema.pattern !== "string") throw new PayloadError(`pattern at ${path} must be a string`);
    try {
      new RegExp(schema.pattern);
    } catch {
      throw new PayloadError(`invalid pattern at ${path}`);
    }
  }
  if (Object.hasOwn(schema, "enum") && !Array.isArray(schema.enum)) {
    throw new PayloadError(`enum at ${path} must be an array`);
  }
  if (Object.hasOwn(schema, "type")) {
    const types = Array.isArray(schema.type) ? schema.type : [schema.type];
    if (types.length === 0 || !types.every((t) => typeof t === "string" && TYPES.has(t))) {
      throw new PayloadError(`unsupported type keyword value at ${path}`);
    }
  }
}

function validateAgainstSchema(payload, schema, path = "$payload") {
  assertSupportedSchema(schema);

  if (Object.hasOwn(schema, "type")) {
    const types = Array.isArray(schema.type) ? schema.type : [schema.type];
    const checks = {
      object: (v) => isObject(v),
      string: (v) => typeof v === "string",
      integer: (v) => typeof v === "number" && Number.isSafeInteger(v),
      array: (v) => Array.isArray(v),
      null: (v) => v === null,
    };
    if (!types.some((t) => checks[t](payload))) fail(`must be ${types.join(" or ")}`, path);
  }

  if (Object.hasOwn(schema, "const") && !sameJsonValue(payload, schema.const)) {
    fail(`must equal ${JSON.stringify(schema.const)}`, path);
  }
  if (Object.hasOwn(schema, "enum") && !schema.enum.some((candidate) => sameJsonValue(payload, candidate))) {
    fail("is not an allowed value", path);
  }
  if (Object.hasOwn(schema, "pattern") && typeof payload === "string" &&
      !new RegExp(schema.pattern).test(payload)) {
    fail("does not match the required pattern", path);
  }
  // JSON Schema length is in Unicode code points, not UTF-16 code units.
  const codePointLength = typeof payload === "string" ? [...payload].length : null;
  if (Object.hasOwn(schema, "minLength") && codePointLength !== null && codePointLength < schema.minLength) {
    fail(`must contain at least ${schema.minLength} code points`, path);
  }
  if (Object.hasOwn(schema, "maxLength") && codePointLength !== null && codePointLength > schema.maxLength) {
    fail(`must contain at most ${schema.maxLength} code points`, path);
  }
  if (Object.hasOwn(schema, "minimum") && typeof payload === "number" && payload < schema.minimum) {
    fail(`must be at least ${schema.minimum}`, path);
  }
  if (Object.hasOwn(schema, "maximum") && typeof payload === "number" && payload > schema.maximum) {
    fail(`must be at most ${schema.maximum}`, path);
  }
  if (Object.hasOwn(schema, "maxItems") && Array.isArray(payload) && payload.length > schema.maxItems) {
    fail(`must contain at most ${schema.maxItems} items`, path);
  }
  if (Object.hasOwn(schema, "items") && Array.isArray(payload)) {
    payload.forEach((item, i) => validateAgainstSchema(item, schema.items, `${path}[${i}]`));
  }

  if (Object.hasOwn(schema, "required")) {
    if (!isObject(payload)) fail("must be an object to satisfy required fields", path);
    for (const name of schema.required) {
      if (!Object.hasOwn(payload, name)) fail(`is missing required property ${JSON.stringify(name)}`, path);
    }
  }
  if (Object.hasOwn(schema, "properties")) {
    if (!isObject(payload)) fail("must be an object to satisfy properties", path);
    for (const [name, value] of Object.entries(payload)) {
      if (!Object.hasOwn(schema.properties, name)) {
        if (schema.additionalProperties === false) fail(`contains unsupported property ${JSON.stringify(name)}`, path);
        continue;
      }
      validateAgainstSchema(value, schema.properties[name], `${path}.${name}`);
    }
  }
  if (schema.additionalProperties === false && !Object.hasOwn(schema, "properties")) {
    if (!isObject(payload)) fail("must be an object for additionalProperties check", path);
    fail("additionalProperties false requires properties", path);
  }
}

function loadSchema(kind, pv = 1) {
  if (!Object.hasOwn(SCHEMAS, kind)) {
    throw new PayloadError(`unsupported payload kind ${JSON.stringify(kind)}`);
  }
  let schema = SCHEMAS[kind];
  if (!Object.hasOwn(schema, "$schema")) {
    // A pv-keyed family: pick the version the payload names.
    if (!Number.isSafeInteger(pv) || !Object.hasOwn(schema, String(pv))) {
      throw new PayloadError(`unsupported pv ${JSON.stringify(pv)} for payload kind ${JSON.stringify(kind)}`);
    }
    schema = schema[String(pv)];
  }
  assertSupportedSchema(schema);
  return schema;
}

/** Validate one parsed payload; reject unknown schema keywords before checking it. */
export function validatePayload(kind, payload) {
  try {
    canonicalize(payload);
  } catch (error) {
    throw new PayloadError(`invalid canonical payload: ${error.message}`);
  }
  validateAgainstSchema(payload, loadSchema(kind, isObject(payload) ? payload.pv : undefined));
  return payload;
}

/** Validate raw JSON text so duplicate keys and floats are rejected pre-parse. */
export function validatePayloadJson(kind, jsonText) {
  if (typeof jsonText !== "string") throw new PayloadError("payload JSON must be text");
  let payload;
  try {
    canonicalizeJson(jsonText);
    payload = JSON.parse(jsonText);
  } catch (error) {
    throw new PayloadError(`invalid canonical payload JSON: ${error.message}`);
  }
  return validatePayload(kind, payload);
}

// Exported for focused fail-closed tests; product callers should use validatePayload.
export function validateSchema(payload, schema) {
  canonicalize(payload);
  validateAgainstSchema(payload, schema);
  return payload;
}
